### ISSUE TITLE
[Bug]: Qwen3-Omni's Audio is interrupted and no sound is output until reach max length.

### ISSUE THREAD

**Issue by yinpeiqi** [vllm-project/vllm-omni#494]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

I am not sure whether bug is caused by the Qwen3-Omni model itself or from our framework.

When the inputs contains an Image, the model may generate long text output (i.e., about 200 text tokens). I test on ``food101`` image dataset, the input prompt is ``Describe this image.``. the output text look likes:
```
Prompt: Describe this image.

Response: Of course, here is a detailed description of the image provided.

This is a close-up photograph of a dessert plate, likely served in a restaurant or café. The lighting is dim, suggesting an evening setting, which makes the food stand out as the central focus.

The main components of the dish are:

*   **Fried Dough:** A serving of what appears to be fried dough, possibly beignets or churros, is presented on a white rectangular plate. They are golden-brown, have a slightly rough texture, and are generously dusted with powdered sugar. Some of the sugary coating has fallen onto the plate.
*   **Dipping Sauces:** There are two small, white ceramic bowls containing dipping sauces.
    *   One bowl contains a dark, glossy sauce, likely a berry compote such as raspberry or blackberry.
    *   The other bowl holds a lighter-colored, creamy sauce that could be a vanilla custard, whipped cream, or a sweetened sour cream.
*   **Presentation:** The entire arrangement rests on a delicate, white paper doily with a lace-like pattern, adding a touch of elegance to the presentation. In the background, a portion of another plate and a fork can be seen, further indicating a dining context.
```

For testing on ``food101`` dataset, some test cases run normally. However, some tests may results in wrong audio output. The sound could normally generate at the begining, but suddenly stopped at middle (i.e., maybe after 30 seconds, ~400 iterations). Next, the model continue to run for the rest of the iteration without normal audio; until reach the max generation lengths (4096 here). Here is an example wrong output:

[sample_2_audio.wav](https://github.com/user-attachments/files/24352847/sample_2_audio.wav)

I found that the audio stop when reach the second paragraph which starts with ``*   **Dipping Sauces:**``.

I print the talker output code, I found that after ~400 iterations, the generated code are quite similar during the slience generation:
```
...
(Worker pid=4005366) code_predictor_codes:  tensor([[1316,  243, 1178,  142,  481, 1030,  825, 1648, 1210, 1370, 1028,  439,                                 
(Worker pid=4005366)          1343, 1430,  802,  241]])                                                                                                      
(Worker pid=4005366) iter: 4084
(Worker pid=4005366) code_predictor_codes:  tensor([[1316,  243, 1178,  142, 1736, 1030, 1978, 1648, 1210, 1370, 1028, 1190,                                 
(Worker pid=4005366)          1343,  650,  802,  851]])
(Worker pid=4005366) iter: 4085
(Worker pid=4005366) code_predictor_codes:  tensor([[1316,  243, 1178,  142,  481, 1030,  825, 1648, 1437, 1370,  662, 1190,
(Worker pid=4005366)          1123,  650,  144,  667]])
(Worker pid=4005366) iter: 4086                                                        
(Worker pid=4005366) code_predictor_codes:  tensor([[1316,  243, 1178,  142,  481, 1030,  825, 1648, 1437, 1370,  662, 1190,
(Worker pid=4005366)          1123,  650,  144,  667]])
(Worker pid=4005366) iter: 4087
...
```

Next, I try to use a new prompt: ``Describe this image. Don't output any '*' in the text.``, and by this mean the generated audio seems normal.

This time, the output text is:
```
Prompt: Describe this image. Don't output any '*' in the text.

Response: This is a close-up photograph of a dessert plate, likely served at a restaurant or café.

The main subject is a serving of what appears to be fried dough, possibly beignets or fritters. They are golden-brown, coated in powdered sugar, and presented in a small paper cone with "ZIN" printed on it in red lettering.

The fried dough rests on a square white ceramic plate, which is adorned with a delicate white paper doily underneath the items.

Accompanying the fried dough are three small, round white bowls containing dipping sauces:
*   One bowl contains a dark, thick sauce, such as chocolate or berry coulis.
*   Another bowl holds a lighter-colored, creamy sauce, like vanilla custard or whipped cream.
*   The third bowl contains a reddish-brown sauce, possibly caramel or a fruit compote.

The background is dark and out of focus, suggesting an indoor setting with dim lighting. A fork is partially visible on a separate plate in the upper left corner.
```

The output audio is:

[sample_2_audio_correct.wav](https://github.com/user-attachments/files/24352867/sample_2_audio_correct.wav)

The script is:
```
python reproduce_bug.py --num_sample 2 --image
```

The code to reproduce the issue (I use greedy sampling and fix the random seed):

```python
#!/usr/bin/env python3
"""
Unified Workload Test for Qwen3-Omni (vLLM)
Combines Audio, Video, and Visual tests using vllm-omni.
"""
import os
import json
import sys
import time
import argparse
import soundfile as sf
import numpy as np
import torch
import random
from PIL import Image
from transformers import set_seed

def seed_everything(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    set_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

def get_omni_model(model_path, workspace_dir, args):

    try:
        from vllm_omni.entrypoints.omni import Omni
    except ImportError:
        print("Error: Could not import vllm_omni. Make sure the path is correct.")
        sys.exit(1)

    log_subdir = "visual_test"
    
    log_file = os.path.join(workspace_dir, f"omni_test/{log_subdir}/results/vllm_stats")
    # Ensure directory exists
    os.makedirs(os.path.dirname(log_file), exist_ok=True)

    print(f"Initializing Omni model from {model_path}...")
    omni_llm = Omni(
        model=model_path,
        stage_configs_path=None,
        log_file=log_file,
        log_stats=True,
    )
    return omni_llm

def get_sampling_params():
    from vllm import SamplingParams
    SEED = 42
    
    thinker_sampling_params = SamplingParams(
        temperature=0.0,
        top_p=1.0,
        top_k=-1,
        max_tokens=1200,
        repetition_penalty=1.05,
        logit_bias={},
        seed=SEED,
    )

    talker_sampling_params = SamplingParams(
        temperature=0.0,
        top_p=1.0,
        top_k=-1,
        max_tokens=4096,
        seed=SEED,
        detokenize=False,
        repetition_penalty=1.05,
        stop_token_ids=[2150],  # TALKER_CODEC_EOS_TOKEN_ID
    )

    code2wav_sampling_params = SamplingParams(
        temperature=0.0,
        top_p=1.0,
        top_k=-1,
        max_tokens=4096 * 16,
        seed=SEED,
        detokenize=True,
        repetition_penalty=1.1,
    )

    return [
        thinker_sampling_params,
        talker_sampling_params,
        code2wav_sampling_params,
    ]

def run_visual_test_vllm(omni_llm, sampling_params_list, num_samples, workspace_dir, args):
    from datasets import load_dataset
    DEFAULT_VISUAL_PROMPT = "Describe this image."
    # DEFAULT_VISUAL_PROMPT = "Describe this image. Don't output any '*' in the text."

    print("\n" + "=" * 60)
    print("Running Visual Workload Test (vLLM)")
    print("=" * 60)
    
    test_dir = os.path.join(workspace_dir, "omni_test/visual_test")
    data_dir = os.path.join(test_dir, "data")
    output_dir = os.path.join(test_dir, "outputs_vllm")
    results_dir = os.path.join(test_dir, "results")
    
    os.makedirs(data_dir, exist_ok=True)
    os.makedirs(output_dir, exist_ok=True)
    os.makedirs(results_dir, exist_ok=True)
    
    print(f"Loading food101 dataset (first {num_samples} samples)...")
    dataset = load_dataset("food101", split=f"validation[:{num_samples}]")
    
    default_system = (
        "You are Qwen, a virtual human developed by the Qwen Team, Alibaba "
        "Group, capable of perceiving auditory and visual inputs, as well as "
        "generating text and speech."
    )
    
    prompts = []
    sample_ids = []
    reference_captions = []

    print("Preparing inputs...")
    for i in range(min(len(dataset), num_samples)):
        sample = dataset[i]
        image = sample["image"]
        image_path = os.path.join(data_dir, f"sample_{i+1}.jpg")
        if not os.path.exists(image_path):
            image.save(image_path)
        
        # Ensure RGB
        if image.mode != "RGB":
            image = image.convert("RGB")

        prompt_text = DEFAULT_VISUAL_PROMPT
        prompt_str = (
            f"<|im_start|>system\n{default_system}<|im_end|>\n"
            "<|im_start|>user\n<|vision_start|><|image_pad|><|vision_end|>"
            f"{prompt_text}<|im_end|>\n"
            f"<|im_start|>assistant\n"
        )
        
        inputs = {
            "prompt": prompt_str,
            "multi_modal_data": {
                "image": image,
            },
        }
        prompts.append(inputs)
        sample_ids.append(i+1)
        
        reference_captions.append(str(sample.get('label', '')))

    print(f"Prepared {len(prompts)} prompts.")
    
    print("\nGenerating responses (Batch)...")
    start_time = time.time()
    omni_outputs = omni_llm.generate(prompts, sampling_params_list)
    total_time = time.time() - start_time
    print(f"Generation completed in {total_time:.2f}s")
    
    results = []
    for idx, s_id in enumerate(sample_ids):
        results.append({
            "sample_id": s_id,
            "prompt": DEFAULT_VISUAL_PROMPT,
            "text_output": "",
            "has_audio_output": False,
            "reference_caption": reference_captions[idx]
        })
        
    avg_time = total_time / len(prompts) if prompts else 0
    
    print("Processing outputs...")
    for stage_outputs in omni_outputs:
        output_type = stage_outputs.final_output_type
        
        for output in stage_outputs.request_output:
            req_id = int(output.request_id.split("_")[0])
            if req_id >= len(results):
                continue
            
            result = results[req_id]
            
            if output_type == "text":
                text_output = output.outputs[0].text
                result["text_output"] = text_output
                
                # Save text
                text_output_path = os.path.join(output_dir, f"sample_{result['sample_id']}_text.txt")
                with open(text_output_path, "w", encoding="utf-8") as f:
                    f.write(f"Prompt: {DEFAULT_VISUAL_PROMPT}\n\n")
                    if result.get("reference_caption"):
                        f.write(f"Reference Caption: {result['reference_caption']}\n\n")
                    f.write(f"Response: {text_output}\n")
                    
            elif output_type == "audio":
                if "audio" in output.multimodal_output:
                    audio_tensor = output.multimodal_output["audio"]
                    audio_numpy = audio_tensor.float().detach().cpu().numpy()
                    if audio_numpy.ndim > 1:
                        audio_numpy = audio_numpy.flatten()
                        
                    audio_output_path = os.path.join(output_dir, f"sample_{result['sample_id']}_audio.wav")
                    sf.write(audio_output_path, audio_numpy, samplerate=24000, format="WAV")
                    result["has_audio_output"] = True

    # Save results
    for r in results:
        r["output_decode_time"] = avg_time
        r["total_time_batch"] = total_time

    result_json_path = os.path.join(results_dir, "Qwen3-Omni_vllm_image.json")
    with open(result_json_path, "w", encoding="utf-8") as f:
        json.dump(results, f, ensure_ascii=False, indent=4)
    print(f"Results saved to {result_json_path}")

def main():
    parser = argparse.ArgumentParser(description="Run Qwen3-Omni Unified Workload Test (vLLM)")
    parser.add_argument("--image", action="store_true", help="Run visual test")
    parser.add_argument("--num_samples", type=int, default=10, help="Number of samples to run")
    parser.add_argument("--model_path", type=str, default="Qwen/Qwen3-Omni-30B-A3B-Instruct", help="Path to model")
    
    args = parser.parse_args()

    seed_everything(42)
    workspace_dir = "/workspace/pqbuild"
    
    # Initialize Omni Model
    omni_llm = get_omni_model(args.model_path, workspace_dir, args)
    sampling_params_list = get_sampling_params()

    if args.image:
        run_visual_test_vllm(omni_llm, sampling_params_list, args.num_samples, workspace_dir, args)

    print("\nAll requested tests completed!")

if __name__ == "__main__":
    main()
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by kechengliu97** [comment]
Please assign this issue to me @hsliuustc0106

**Comment by yenuo26** [comment]
I also encountered this issue, but the phenomenon is slightly different. My text is: I cannot process or analyze audio, images, or videos directly. If you have specific questions about the content of an audio file, image, or video, please provide a detailed description or context, and I will do my best to assist you based on that information.
But, My audio stops playing after outputting "I cannot process or analyze audio, images, or videos directly. If you have specific questions about the content of an audio file"

**Comment by hsliuustc0106** [comment]
> I also encountered this issue, but the phenomenon is slightly different. My text is: I cannot process or analyze audio, images, or videos directly. If you have specific questions about the content of an audio file, image, or video, please provide a detailed description or context, and I will do my best to assist you based on that information. But, My audio stops playing after outputting "I cannot process or analyze audio, images, or videos directly. If you have specific questions about the content of an audio file"

please open a new issue and link this issue

**Comment by kechengliu97** [comment]
Using the same figure, prompt as input, I get the outputs as shown below.

# Transformer Output in Text
```
"Of course, here is a detailed description of the image provided.\n\n

This is a close-up photograph of a meal, likely taken at a café or casual restaurant. The scene is set on a dark green tray.\n\n

The main subject is a plate of three beignets, which are a traditional New Orleans pastry. They are square-shaped, deep-fried, and have a golden-brown, slightly puffy appearance. The beignets are heavily dusted with powdered sugar, which also lightly coats the white paper plate they are served on and the tray around them.\n\n

To the right of the plate, there is a white paper cup with a clear plastic lid and a straw. The cup has a logo that reads \"Café Du Monde\" in a stylized font, accompanied by a line drawing of a building. This strongly suggests the beignets are from the famous Café du Monde in New Orleans. The cup appears to contain a dark beverage, most likely coffee.\n\nA white paper straw wrapper is visible on the tray between the beignets and the coffee cup. The overall lighting is somewhat dim, creating a cozy atmosphere."
```

# vLLM-Omni Output in Text

```
Of course, here is a detailed description of the image.

This is a close-up photograph of a classic New Orleans breakfast or snack, featuring beignets and coffee served on a green tray.

## After above paragraphs, the output wav file remains blank for about 5 mins

- **Main Subject (Beignets):** The central focus is a plate piled high with several square-shaped beignets. These deep-fried pastries have a golden-brown, slightly puffy appearance. They are heavily dusted with powdered sugar, which covers their surface and has also fallen onto the white paper plate beneath them and the green tray.

- **Beverage:** To the right of the beignets sits a white disposable cup from "Café Du Monde," a famous coffee shop in New Orleans. The cup features the café's logo, which includes an illustration of its building. The cup is filled with a dark beverage, likely chicory coffee, and is sealed with a clear plastic lid that has a drinking hole. A white plastic straw is placed across the top of the cup.

- **Setting & Composition:** The food and drink are arranged on a dark green plastic tray, typical of a casual eatery. The lighting appears to be natural daylight, casting soft shadows. The overall composition suggests a simple, authentic dining experience at a local café.
```

for vLLM-Omni output wav file, the first 2 paragraphs are read while the rest vanishes. I'm keep on tracking the root cause.

**Comment by hsliuustc0106** [comment]
> Using the same figure, prompt as input, I get the outputs as shown below.
> 
> # Transformer Output in Text
> ```
> "Of course, here is a detailed description of the image provided.\n\n
> 
> This is a close-up photograph of a meal, likely taken at a café or casual restaurant. The scene is set on a dark green tray.\n\n
> 
> The main subject is a plate of three beignets, which are a traditional New Orleans pastry. They are square-shaped, deep-fried, and have a golden-brown, slightly puffy appearance. The beignets are heavily dusted with powdered sugar, which also lightly coats the white paper plate they are served on and the tray around them.\n\n
> 
> To the right of the plate, there is a white paper cup with a clear plastic lid and a straw. The cup has a logo that reads \"Café Du Monde\" in a stylized font, accompanied by a line drawing of a building. This strongly suggests the beignets are from the famous Café du Monde in New Orleans. The cup appears to contain a dark beverage, most likely coffee.\n\nA white paper straw wrapper is visible on the tray between the beignets and the coffee cup. The overall lighting is somewhat dim, creating a cozy atmosphere."
> ```
> 
> # vLLM-Omni Output in Text
> ```
> Of course, here is a detailed description of the image.
> 
> This is a close-up photograph of a classic New Orleans breakfast or snack, featuring beignets and coffee served on a green tray.
> 
> ## After above paragraphs, the output wav file remains blank for about 5 mins
> 
> - **Main Subject (Beignets):** The central focus is a plate piled high with several square-shaped beignets. These deep-fried pastries have a golden-brown, slightly puffy appearance. They are heavily dusted with powdered sugar, which covers their surface and has also fallen onto the white paper plate beneath them and the green tray.
> 
> - **Beverage:** To the right of the beignets sits a white disposable cup from "Café Du Monde," a famous coffee shop in New Orleans. The cup features the café's logo, which includes an illustration of its building. The cup is filled with a dark beverage, likely chicory coffee, and is sealed with a clear plastic lid that has a drinking hole. A white plastic straw is placed across the top of the cup.
> 
> - **Setting & Composition:** The food and drink are arranged on a dark green plastic tray, typical of a casual eatery. The lighting appears to be natural daylight, casting soft shadows. The overall composition suggests a simple, authentic dining experience at a local café.
> ```
> 
> for vLLM-Omni output wav file, the first 2 paragraphs are read while the rest vanishes. I'm keep on tracking the root cause.

we are approaching the hidden bugs

**Comment by kechengliu97** [comment]
After printing the Talker input passed to the Code2Wav worker, I found many repeated prompt token ID segments such as [...2029, 2029, 2029, 2029, 2029, 2029, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316, 1316...]. This is likely the root cause of the long blank segments in the audio output from Code2Wav. Upon revisiting the test script, I noticed that the Temperature value in Talker’s SamplingParams was set to 0.0. After changing it to 0.9, the audio output returned to normal. It can be concluded that the issue was caused by setting the Temperature in Talker’s SamplingParams to 0.0.

**Comment by kechengliu97** [comment]
An attempt was made to intercept the Temperature at the Talker stage, but it was unsuccessful. The task is now reassigned to @tzhouam . Please have Taichang Zhou carefully review the parameter transmission path and implement the interception correctly.

**Comment by hsliuustc0106** [comment]
closed since #688 merged
